source/fab/entry.py

```diff
@@ -0,0 +1,105 @@
+                                     description=description)
+    parser.add_argument('-h', '-help', '--help', action='help',
+                        help='Print this help and exit')
+    parser.add_argument('-V', '--version', action='version',
+                        version=fab.__version__,
+                        help='Print version identifier and exit')
+    parser.add_argument('-v', '--verbose', action='store_true',
+                        help='Produce a running commentary on progress')
+    parser.add_argument('-w', '--workspace', metavar='FILENAME', type=Path,
+                        help='Directory for working files.')
+    # TODO: Details like these flags will eventually come from
+    #       our configuration system
+    parser.add_argument('--fpp-flags', action='store', type=str, default='',
+                        help='Provide flags for Fortran PreProcessor ')
+    parser.add_argument('source', type=Path,
+                        help='The path of the source tree to build')
+    return parser.parse_args()
+
+
+def fab_entry() -> None:
+    """
+    The core Fab build tool.
+    """
+    logger = logging.getLogger('fab')
+    logger.addHandler(logging.StreamHandler(sys.stderr))
+
+    arguments = fab_cli()
+
+    if arguments.verbose:
+        logger.setLevel(logging.INFO)
+    else:
+        logger.setLevel(logging.WARNING)
+
+    if not arguments.workspace:
+        arguments.workspace = arguments.source / 'working'
+
+    application = fab.application.Fab(arguments.workspace,
+                                      arguments.fpp_flags)
+    application.run(arguments.source)
+
+
+def dump_cli() -> argparse.Namespace:
+    """
+    Parse command line arguments for the dumper tool.
+    """
+    description = 'Flexible build system for scientific software.'
+    parser = argparse.ArgumentParser(add_help=False,
+                                     description=description)
+    parser.add_argument('-h', '-help', '--help', action='help',
+                        help='Print this help and exit')
+    parser.add_argument('-V', '--version', action='version',
+                        version=fab.__version__,
+                        help='Print version identifier and exit')
+    parser.add_argument('-v', '--verbose', action='store_true',
+                        help='Produce a running commentary on progress')
+    parser.add_argument('-w', '--workspace', metavar='FILENAME', type=Path,
+                        help='Directory for working files.')
+    return parser.parse_args()
+
+
+def dump_entry() -> None:
+    """
+    Dump a state database from a working directory.
+    """
+    logger = logging.getLogger('fab-dumper')
+    logger.addHandler(logging.StreamHandler(sys.stderr))
+
+    arguments = dump_cli()
+
+    if arguments.verbose:
+        logger.setLevel(logging.INFO)
+    else:
+        logger.setLevel(logging.WARNING)
+
+    if not arguments.workspace:
+        arguments.workspace = Path.cwd() / 'working'
```

Comment:
Should we be bothered that the location of the default working directory is subtly different between the two tools?  I understand that the dumper tool cannot use the source directory (as there's no such concept) but perhaps the default location for the main tool should be in `cwd` also rather than in `source`?  This also ties into the fact that perhaps we don't want to be adding anything to the source directory if we can help it.  

Could do this as a followup though as it's not really in scope

Reply:
I take your points about divergant behaviour and not putting the working directory in the source directory. I'm not sure what the correct default behaviour should be. I think postponing this to a different ticket is the way to go. See #43.